service/s3/handwritten_paginators_test.go

```diff
@@ -0,0 +1,160 @@
+package s3
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/smithy-go/middleware"
+	"testing"
+)
+
+var bucket *string
+var limit int32
+var keyMarker *string
+var idMarker *string
+
+type testListMPUMiddleware struct {
+	id int
+}
+
+type testListOVMiddleware struct {
+	id int
+}
+
+func (m *testListMPUMiddleware) ID() string {
+	return fmt.Sprintf("mock middleware %d", m.id)
+}
+
+func (m *testListMPUMiddleware) HandleInitialize(ctx context.Context, input middleware.InitializeInput, next middleware.InitializeHandler) (
```

Comment:
**fix**: This shouldn't be necessary. The paginator takes an interface for the client type:

```go
type ListObjectVersionsAPIClient interface {
	ListObjectVersions(context.Context, *ListObjectVersionsInput, ...func(*Options)) (*ListObjectVersionsOutput, error)
}
```

You can just implement a test/mock version of this interface which removes the need for middleware or constructing a real client and allows you to directly control the outputs and observe/record the inputs to make assertions.

e.g.

```go
// this is just an example
type mockListObjectVersionsClient struct {
    responses []*ListObjectVersionsOutput
    inputs []*ListObjectVersionsOutput
    requestCount int
}

func (c *mockListObjectVersionsClient) ListObjectVersions(ctx context.Context, input *ListObjectVersionsInput, options ...func(*Options)) (*ListObjectVersionsOutput, error) {
    idx := c.requestCount + 1
    if idx > len(c.resopnses) {   // error }
    c.requestCount = idx     
    // record the input observed
    c.inputs = append(c.inputs, input)
    return c.responses[idx], nil
}
```